File: gemini/nivel avanzado/java/src/Palindromo.java
import java.util.Scanner;
import java.text.Normalizer; // Para manejar tildes y caracteres especiales

public class Palindromo {

    public static class ResultadoPalindromo {
        boolean esPalindromo;
        String mensaje;

        ResultadoPalindromo(boolean esPalindromo, String mensaje) {
            this.esPalindromo = esPalindromo;
            this.mensaje = mensaje;
        }
    }

    public static ResultadoPalindromo esPalindromo(String cadena) {
        if (cadena == null || cadena.trim().isEmpty()) {
            return new ResultadoPalindromo(false, "La cadena original está vacía o solo contiene espacios.");
        }

        // Normalizar para quitar tildes y otros diacríticos (opcional pero recomendado)
        String cadenaNormalizada = Normalizer.normalize(cadena, Normalizer.Form.NFD);
        cadenaNormalizada = cadenaNormalizada.replaceAll("\\p{InCombiningDiacriticalMarks}+", "");

        // Convertir a minúsculas y quitar caracteres no alfanuméricos
        String cadenaLimpia = cadenaNormalizada.toLowerCase().replaceAll("[^a-z0-9]", "");

        if (cadenaLimpia.isEmpty()) {
             return new ResultadoPalindromo(false, "La cadena procesada está vacía (la original solo contenía espacios, puntuación o caracteres no alfanuméricos).");
        }

        int i = 0;
        int j = cadenaLimpia.length() - 1;
        while (i < j) {
            if (cadenaLimpia.charAt(i) != cadenaLimpia.charAt(j)) {
                return new ResultadoPalindromo(false, null);
            }
            i++;
            j--;
        }
        return new ResultadoPalindromo(true, null);
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Ingresa una cadena de texto para verificar si es un palíndromo: ");
        String cadenaOriginal = scanner.nextLine();

        ResultadoPalindromo resultado = esPalindromo(cadenaOriginal);

        if (resultado.mensaje != null) {
            System.out.println(resultado.mensaje);
             if (!resultado.esPalindromo && resultado.mensaje.startsWith("La cadena procesada está vacía")){
                 System.out.println("Por lo tanto, '" + cadenaOriginal + "' NO se considera un palíndromo en este contexto.");
             } else if (!resultado.esPalindromo && resultado.mensaje.startsWith("La cadena original está vacía")) {
                  System.out.println("'" + cadenaOriginal + "' NO es un palíndromo.");
             }
        } else if (resultado.esPalindromo) {
            System.out.println("¡Felicidades! La cadena '" + cadenaOriginal + "' SÍ es un palíndromo.");
        } else {
            System.out.println("La cadena '" + cadenaOriginal + "' NO es un palíndromo.");
        }
        scanner.close();
    }
}

File: gemini/nivel avanzado/ruby/palindromo.rb
# Para manejo de tildes más avanzado, se podría usar una gema como 'unicode_utils'
# require 'unicode_utils/downcase'
# require 'unicode_utils/nfkd_char'

def es_palindromo(cadena)
  return { es_pal: false, mensaje: "La cadena original está vacía o solo contiene espacios.", cadena_limpia: "" } if cadena.nil? || cadena.strip.empty?

  # Normalización simple para quitar tildes comunes (puede no cubrir todos los casos unicode)
  # Para una solución más robusta, se usaría unicode_utils o similar.
  cadena_normalizada = cadena.downcase
                             .gsub(/[áäâàãåāăąǎǻǽ]/, 'a')
                             .gsub(/[éëêèęėēĕəǝǣ]/, 'e')
                             .gsub(/[íïîìįīĭǐ]/, 'i')
                             .gsub(/[óöôòõøōŏőơǒǿ]/, 'o')
                             .gsub(/[úüûùųūŭůűưǔǖǘǚǜ]/, 'u')
                             .gsub(/[ñńņňŉ]/, 'n')
                             .gsub(/[çćčĉċ]/, 'c')
                             .gsub(/[ýÿŷ]/, 'y')
                             .gsub(/[źżž]/, 'z')
                             # quitar otros no alfanuméricos
                             .gsub(/[^a-z0-9]/, '')


  # Alternativa más simple sin manejo de tildes:
  # cadena_limpia = cadena.downcase.gsub(/[^a-z0-9]/, '')
  cadena_limpia = cadena_normalizada # Usamos la normalizada

  return { es_pal: false, mensaje: "La cadena procesada está vacía (la original solo contenía puntuación o caracteres no alfanuméricos).", cadena_limpia: cadena_limpia } if cadena_limpia.empty?
  
  es_resultado = cadena_limpia == cadena_limpia.reverse
  { es_pal: es_resultado, mensaje: nil, cadena_limpia: cadena_limpia }
end

puts "Ingresa una cadena de texto para verificar si es un palíndromo:"
cadena_original = gets.chomp

resultado = es_palindromo(cadena_original)

unless resultado[:cadena_limpia].empty?
    puts "Cadena procesada (sin espacios/puntuación, en minúsculas y tildes simplificadas): '#{resultado[:cadena_limpia]}'"
end


if resultado[:mensaje]
  puts resultado[:mensaje]
  if !resultado[:es_pal] && resultado[:mensaje].include?("procesada está vacía")
    puts "Por lo tanto, '#{cadena_original}' NO se considera un palíndromo en este contexto."
  elsif !resultado[:es_pal] && resultado[:mensaje].include?("original está vacía")
    puts "'#{cadena_original}' NO es un palíndromo."
  end
elsif resultado[:es_pal]
  puts "¡Felicidades! La cadena '#{cadena_original}' SÍ es un palíndromo."
else
  puts "La cadena '#{cadena_original}' NO es un palíndromo."
end